builder/haddock-builder.nix

```diff
@@ -0,0 +1,180 @@
+{ stdenv, buildPackages, lib, haskellLib, ghc, ghcForComponent, nonReinstallablePkgs, runCommand, writeText, writeScript, makeConfigFiles }:
+
+{ componentId
+, component
+, package
+, name
+, setup
+, src
+, flags
+, revision
+, patches
+
+, preUnpack
+, postUnpack
+, configureFlags
+, preConfigure
+, postConfigure
+, preBuild
+, postBuild
+, preHaddock
+, postHaddock
+, setupInstallFlags
+
+, doHoogle
+, hyperlinkSource
+, setupHaddockFlags
+, setupGhcOptions
+
+, needsProfiling
+, componentDrv
+, configFiles # component config files
+}:
+
+let
+  # the target dir for haddock documentation
+  docdir = docoutput: docoutput + "/share/doc/" + componentId.cname;
+
+  packageCfgDir = configFiles.packageCfgDir;
+
+  fullName = if haskellLib.isAll componentId
+    then "${name}-haddock-all"
+    else "${name}-${componentId.ctype}-${componentId.cname}-haddock";
+
+  docsConfigFiles = makeConfigFiles {
+    inherit (package) identifier;
+    inherit component fullName flags needsProfiling;
+    chooseDrv = p: p.haddock;
+  };
+
+  # let
+  #   # depDocConfig = map (x: x.configFiles) (haskellLib.flatLibDepends component);
+  # in runCommand "${ghc.targetPrefix}${fullName}-docs-config" {
+  #   nativeBuildInputs = [ghc];
+  # } (''
+  #   mkdir -p $out
+
+  #   for i in cabal.config configure-flags ghc-environment; do
+  #     substitute "${configFiles}/$i" "$out/$i"  --replace "${configFiles}" "$out"
+  #   done
+
+  #   mkdir -p $out/${packageCfgDir}
+
+  #   # Copy over the nonReinstallablePkgs from the global package db.
+  #   ${lib.concatMapStringsSep "\n" (p: ''
+  #     find ${ghc}/lib/${ghc.name}/package.conf.d -name '${p}*.conf' -exec cp -f {} $out/${packageCfgDir} \;
+  #   '') nonReinstallablePkgs}
+
+  # '');
+
+  finalConfigureFlags = lib.concatStringsSep " " (
+    [ "--prefix=${componentDrv}"
+      "--docdir=${docdir "$doc"}"
+      "${haskellLib.componentTarget componentId}"
+      "$(cat ${docsConfigFiles}/configure-flags)"
+      # GHC
+      "--with-ghc=${ghc.targetPrefix}ghc"
+      "--with-ghc-pkg=${ghc.targetPrefix}ghc-pkg"
+      "--with-hsc2hs=${ghc.targetPrefix}hsc2hs"
+    ]
+      ++ configureFlags
+      ++ (ghc.extraConfigureFlags or [])
+    );
+
+  shellWrappers = ghcForComponent {
+    componentName = fullName;
+    configFiles = docsConfigFiles;
+  };
+
+in stdenv.lib.fix (drv: stdenv.mkDerivation ({
+  name = "${ghc.targetPrefix}${fullName}";
+  inherit src;
+
+  LANG = "en_US.UTF-8";         # GHC needs the locale configured during the Haddock phase.
+  LC_ALL = "en_US.UTF-8";
+
+  passthru = {
+    configFiles = docsConfigFiles;
+
+    # The directory containing the haddock documentation.
+    # `null' if no haddock documentation was built.
+    haddockDir = "${docdir drv.doc}/html";
+  };
+
+  outputs = ["out" "doc"];
```

Comment:
Why does the haddock derivation have split outputs?
It could use a comment to explain.

Reply:
Added comments